pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,698 @@
+	source, _, err := getVolumeSource(spec)
+	if err != nil {
+		glog.Warningf("getVolumeSource err: %v", err)
+		return nil, err
+	}
+
+	var apiCfg *storageosAPIConfig
+	if source.SecretRef != nil {
+		if apiCfg, err = parsePodSecret(pod, source.SecretRef.Name, plugin.host.GetKubeClient()); err != nil {
+			glog.Errorf("Couldn't get secret from %v/%v", pod.Namespace, source.SecretRef)
+			return nil, err
+		}
+	}
+
+	return plugin.newMounterInternal(spec, pod, apiCfg, &storageosUtil{}, plugin.host.GetMounter())
+}
+
+func (plugin *storageosPlugin) newMounterInternal(spec *volume.Spec, pod *v1.Pod, apiCfg *storageosAPIConfig, manager storageosManager, mounter mount.Interface) (volume.Mounter, error) {
+	source, readOnly, err := getVolumeSource(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageosMounter{
+		storageos: &storageos{
+			podUID:          pod.UID,
+			podNamespace:    pod.GetNamespace(),
+			pvName:          spec.Name(),
+			volName:         source.VolumeName,
+			volNamespace:    source.VolumeNamespace,
+			fsType:          source.FSType,
+			readOnly:        readOnly,
+			apiCfg:          apiCfg,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(pod.UID, source.VolumeNamespace, source.VolumeName, spec.Name(), plugin.host)),
+		},
+		devicePath:  storageosDevicePath,
+		diskMounter: &mount.SafeFormatAndMount{Interface: mounter, Runner: exec.New()},
+	}, nil
+}
+
+func (plugin *storageosPlugin) NewUnmounter(pvName string, podUID types.UID) (volume.Unmounter, error) {
+	return plugin.newUnmounterInternal(pvName, podUID, &storageosUtil{}, plugin.host.GetMounter())
+}
+
+func (plugin *storageosPlugin) newUnmounterInternal(pvName string, podUID types.UID, manager storageosManager, mounter mount.Interface) (volume.Unmounter, error) {
+
+	// Parse volume namespace & name from mountpoint if mounted
+	volNamespace, volName, err := getVolumeInfo(pvName, podUID, plugin.host)
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageosUnmounter{
+		storageos: &storageos{
+			podUID:          podUID,
+			pvName:          pvName,
+			volName:         volName,
+			volNamespace:    volNamespace,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, volNamespace, volName, pvName, plugin.host)),
+		},
+	}, nil
+}
+
+func (plugin *storageosPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.StorageOS == nil {
+		return nil, fmt.Errorf("spec.PersistentVolumeSource.StorageOS is nil")
+	}
+
+	class, err := util.GetClassForVolume(plugin.host.GetKubeClient(), spec.PersistentVolume)
+	if err != nil {
+		return nil, err
+	}
+	adminSecretName := ""
+	adminSecretNamespace := "default"
```

Comment:
I would not expect a fallback to a default namespace. Either specify the namespace in the storage class, or use the namespace of the PVC, or don't fetch a secret